src/lib/comms/sol-coap-transport-ip.c

```diff
@@ -0,0 +1,310 @@
+/*
+ * This file is part of the Soletta Project
+ *
+ * Copyright (C) 2016 Intel Corporation. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in
+ *     the documentation and/or other materials provided with the
+ *     distribution.
+ *   * Neither the name of Intel Corporation nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+
+#define SOL_LOG_DOMAIN &_log_domain
+#include "sol-log-internal.h"
+#include "sol-coap-transport.h"
+#include "sol-coap-transport-ip.h"
+#include "sol-network.h"
+#include "sol-socket.h"
+#include "sol-util-internal.h"
+
+SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "coap-transport-ip");
+
+#define IPV4_ALL_COAP_NODES_GROUP "224.0.1.187"
+
+#define IPV6_ALL_COAP_NODES_SCOPE_LOCAL "ff02::fd"
+#define IPV6_ALL_COAP_NODES_SCOPE_SITE "ff05::fd"
+
+struct sol_coap_transport_ip {
+    struct sol_coap_transport base;
+    struct sol_socket *sock;
+    struct {
+        bool (*cb)(void *data, struct sol_coap_transport *transport);
+        const void *cb_data;
+    } on_can_read;
+
+    struct {
+        bool (*cb)(void *data, struct sol_coap_transport *transport);
+        const void *cb_data;
+    } on_can_write;
+};
+
+static bool
+transport_ip_on_can_read(void *data, struct sol_socket *s)
+{
+    struct sol_coap_transport_ip *tip = data;
+
+    if (tip->on_can_read.cb)
+        return tip->on_can_read.cb((void *)tip->on_can_read.cb_data, data);
```

Comment:
to avoid confusion and allow compiler to alert on changes, instead of passing the "data" as transport, make it explicit `&tip->base`